cluster.go

```diff
@@ -0,0 +1,464 @@
+	addrs, err := opt.getAddrs()
+	if err != nil {
+		return nil, err
+	}
+
+	client := &ClusterClient{
+		addrs:   addrs,
+		conns:   newLRUPool(opt),
+		opt:     opt,
+		_reload: 1,
+	}
+	client.commandable.process = client.process
+	client.reloadIfDue()
+	return client, nil
+}
+
+// Closes the pool
+func (c *ClusterClient) Close() error {
+	c.cachemx.Lock()
+	defer c.cachemx.Unlock()
+
+	return c.reset()
+}
+
+// GetMasterAddrBySlot finds the current master address for a given hash slot
+func (c *ClusterClient) GetMasterAddrBySlot(hashSlot int) string {
+	if addrs := c.slots[hashSlot]; len(addrs) > 0 {
+		return addrs[0]
+	}
+	return ""
+}
+
+// GetSlaveAddrsBySlot finds the current slave addresses for a given hash slot
+func (c *ClusterClient) GetSlaveAddrsBySlot(hashSlot int) []string {
+	if addrs := c.slots[hashSlot]; len(addrs) > 0 {
+		return addrs[1:]
+	}
+	return []string{}
+}
+
+// GetNodeClientByAddr returns a node's client for a given address
+func (c *ClusterClient) GetNodeClientByAddr(addr string) *Client {
+	return c.conns.Fetch(addr)
+}
+
+// ------------------------------------------------------------------------
+
+func (c *ClusterClient) process(cmd Cmder) {
+	var hashSlot int
+
+	c.reloadIfDue()
+	ask := false
+
+	key := cmd.firstKey()
+	if key != "" {
+		hashSlot = HashSlot(key)
+	} else {
+		hashSlot = rand.Intn(HashSlots)
+	}
+
+	c.cachemx.RLock()
+	defer c.cachemx.RUnlock()
+
+	tried := make(map[string]struct{}, len(c.addrs))
+	addr := c.GetMasterAddrBySlot(hashSlot)
+	for attempt := 0; attempt < c.opt.getMaxRedirects(); attempt++ {
+		tried[addr] = struct{}{}
+
+		// Pick the connection, process request
+		conn := c.GetNodeClientByAddr(addr)
+		if ask {
+			pipe := conn.Pipeline()
+			pipe.Process(NewCmd("ASKING"))
+			pipe.Process(cmd)
+			_, _ = pipe.Exec()
+			ask = false
+		} else {
+			conn.Process(cmd)
+		}
+
+		// If there is no (real) error, we are done!
+		err := cmd.Err()
+		if err == nil || err == Nil {
+			return
+		}
+
+		// On connection errors, pick the next (not previosuly) tried connection
+		// and try again
+		if _, ok := err.(*net.OpError); ok || err == io.EOF {
+			if addr = c.next(tried); addr == "" {
+				return
+			}
+			cmd.reset()
+			continue
+		}
+
+		// Check the error message, return if unexpected
+		parts := strings.SplitN(err.Error(), " ", 3)
+		if len(parts) != 3 {
+			return
+		}
+
+		// Handle MOVE and ASK redirections, return on any other error
+		switch parts[0] {
+		case "MOVED":
+			c.forceReload()
+			addr = parts[2]
+		case "ASK":
+			ask = true
+			addr = parts[2]
+		default:
+			return
+		}
+		cmd.reset()
+	}
+}
+
+// Closes all connections and reloads slot cache, if due
+func (c *ClusterClient) reloadIfDue() (err error) {
+	if !atomic.CompareAndSwapUint32(&c._reload, 1, 0) {
+		return
+	}
+
+	var infos []ClusterSlotInfo
+
+	c.cachemx.Lock()
+	defer c.cachemx.Unlock()
+
+	for _, addr := range c.addrs {
+		c.reset()
```

Comment:
It resets the cache on each attempt https://github.com/antirez/redis-rb-cluster/blob/master/cluster.rb#L60 and then breaks on the first success https://github.com/antirez/redis-rb-cluster/blob/master/cluster.rb#L83